mathesar/api/exceptions/database_exceptions/exceptions.py

```diff
@@ -234,3 +234,77 @@ def __init__(
             'column_id': column.id
         }
         super().__init__(exception, self.error_code, message_str, field, details, status_code)
+
+
+class TypeMismatchViolationAPIException(MathesarAPIException):
+    """ Exception raised when trying to insert a non castable datatype value to a column with certain datatype"""
+    error_code = ErrorCodes.TypeMismatchViolation.value
+
+    def __init__(
+            self,
+            exception,
+            message=None,
+            field=None,
+            details=None,
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
```

Comment:
I think all of these error codes would be more appropriate as status code 400, since the request shouldn't be repeated in the same form (you'll just get the same violation again). 500 is more appropriate for "maybe trying again in a minute will work". @silentninja may have an opinion on this, though.

Reply:
I was also curious about that, as all the exceptions in this file raise 500s and are then overridden to raise a 400 when the actual error/exception occurs.

Reply:
Most of the error classes are written as `Internal Server Error` because 
- Error classes are meant to be reused when the same error can happen in multiple places
- It is assumed the actual reason is due to some bug with the program and not due to the request sent by the user.

We then convert it to a 400 error when we find out the actual reason is due to an incorrect request sent by the user. If an exception is always known to occur due to incorrect data sent by the user, setting the default status code as `400` makes sense.

Reply:
@silentninja That Makes sense, does that mean that `NotNullImportViolationAPIException` and `UniqueImportViolationAPIException` should produce an error code 400 as they are known to get raised only during import, and the rest of the exceptions should produce a 500 as they are more general and may occur at different places?

Reply:
Wouldn't it be better to use the existing `mathesar.api.exceptions.database_exceptions.exceptions.NotNullViolationAPIException` instead of `NotNullImportViolationAPIException`. Both seem to be pointing to the same postgres Exception

Reply:
> Wouldn't it be better to use the existing `mathesar.api.exceptions.database_exceptions.exceptions.NotNullViolationAPIException` instead of `NotNullImportViolationAPIException`. Both seem to be pointing to the same postgres Exception

Yes but, the reason I decided to make new exception classes i.e. `NotNullImportViolationAPIException` and `UniqueImportViolationAPIException` because `NotNullViolationAPIException` and `UniqueViolationAPIException` were getting thrown when trying to set a column constraint on a column consisting invalid values. However, in this case, even though the Postgres exceptions are identical the reason/way they occur are different i.e they are made to be thrown when trying to insert invalid values in a column with these constraints already set.

Reply:
> NotNullViolationAPIException and UniqueViolationAPIException were getting thrown when trying to set a column constraint on a column consisting invalid values

That makes sense @Anish9901 but I noticed that `NotNullViolationAPIException` is thrown when creating a new record 
https://github.com/centerofci/mathesar/blob/c03a2b0e197f2329b227b8aa1e96a06417070f5b/mathesar/api/serializers/records.py#L33
So it looks like they are the same type of error, correct me if I am wrong.

Reply:
Um, you are right I based my decision on the docstring of the exception class, I didn't notice the `NotNullViolationAPIException` was getting thrown while adding an invalid record.